service/src/test/java/bio/terra/cbas/dependencies/sam/TestSamService.java

```diff
@@ -0,0 +1,102 @@
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+import bio.terra.common.sam.exception.SamInterruptedException;
+import bio.terra.common.sam.exception.SamUnauthorizedException;
+import org.broadinstitute.dsde.workbench.client.sam.ApiClient;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.broadinstitute.dsde.workbench.client.sam.api.UsersApi;
+import org.broadinstitute.dsde.workbench.client.sam.model.UserStatusInfo;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.stubbing.Answer;
+import org.springframework.http.HttpStatus;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+class TestSamService {
+  private SamService samService;
+
+  private final String tokenValue = "foo-token";
+  private final String expiredTokenValue = "expired-token";
+  private final String tokenCausingInterrupt = "interrupting-cow-moo";
+  private final UserStatusInfo mockUser =
+      new UserStatusInfo()
+          .userEmail("realuser@gmail.com")
+          .userSubjectId("user-id-foo")
+          .enabled(true);
+
+  @BeforeEach
+  void init() throws ApiException {
+    RequestContextHolder.setRequestAttributes(
+        new ServletRequestAttributes(new MockHttpServletRequest()));
+    UsersApi usersApi = mock(UsersApi.class);
+    ApiClient apiClient = mock(ApiClient.class);
+    SamClient samClient = mock(SamClient.class);
+    samService = spy(new SamService(samClient));
+    when(samClient.getApiClient(any())).thenReturn(apiClient);
+    doReturn(usersApi).when(samService).getUsersApi(any());
+    when(usersApi.getUserStatusInfo())
+        .thenAnswer(
+            (Answer<UserStatusInfo>)
+                invocation -> {
+                  String token = samService.getUserToken();
+                  if (token != null && token.equals(tokenValue)) {
+                    return mockUser;
+                  } else if (token != null && token.equals(tokenCausingInterrupt)) {
+                    throw new InterruptedException();
+                  } else {
+                    // expired or no token
+                    throw new ApiException(401, "Unauthorized :(");
+                  }
+                });
+  }
+
+  void setTokenValue(String token) {
+    MockHttpServletRequest request = new MockHttpServletRequest();
+    request.setAttribute(ATTRIBUTE_NAME_TOKEN, token);
+    RequestContextHolder.setRequestAttributes(new ServletRequestAttributes(request));
+  }
+
+  @Test
+  void testGetUserToken() {
+    setTokenValue(tokenValue);
+    String userToken = samService.getUserToken();
+    assertEquals(tokenValue, userToken);
+  }
+
+  @Test
+  void testGetSamUserNoToken() {
```

Comment:
I would add a test for the caching too, in particular making sure that two different instances of the same token value uses the cache. (by the way, I think that's another reason for making the getSamUser function take parameters rather than relying on threadlocal magic - it makes these tests a lot easier to write!)

Reply:
I did a lot of digging and configuration changes to try to test the caching and it seems that caching doesn't agree well with spring boot tests. Willing to discuss further tomorrow but I kept running into a brick wall today and I can't seem to find examples in other Terra repos or online.